File: ios/ModelsMeterCore/Sources/ModelsMeterCore/UsageParser.swift
import Foundation

public enum UsageParser {
    private static let fiveHours: Int64 = 18_000
    private static let week: Int64 = 604_800
    private static let month: Int64 = 2_592_000
    /// Free-tier accounts report a single ~30-day Codex window; accept 10–45 days so calendar
    /// months and drifting billing periods still classify while staying clear of the weekly
    /// window's 9-day ceiling.
    private static let monthRange: ClosedRange<Int64> = 864_000 ... 3_888_000

    public static func parse(_ string: String, fetchedAt: Date = Date()) throws -> UsageSnapshot {
        try parse(Data(string.utf8), fetchedAt: fetchedAt)
    }

    public static func parse(_ data: Data, fetchedAt: Date = Date()) throws -> UsageSnapshot {
        let root = try JSONSupport.object(from: data)
        let planType = JSONSupport.string(root["plan_type"])

        var nextID = 0
        func candidate(from object: JSONSupport.Object?) -> Candidate? {
            guard let window = parseWindow(object) else {
                return nil
            }
            defer { nextID += 1 }
            return Candidate(id: nextID, window: window)
        }

        let rateLimit = JSONSupport.object(root["rate_limit"])
        let allowed: Bool
        let limitReached: Bool
        let primary: Candidate?
        let secondary: Candidate?
        var primaryCandidates: [Candidate] = []

        if let rateLimit {
            allowed = JSONSupport.bool(rateLimit["allowed"], default: true)
            limitReached = JSONSupport.bool(rateLimit["limit_reached"], default: false)
            primary = candidate(from: JSONSupport.object(rateLimit["primary_window"]))
            secondary = candidate(from: JSONSupport.object(rateLimit["secondary_window"]))
            if let primary { primaryCandidates.append(primary) }
            if let secondary { primaryCandidates.append(secondary) }
        } else {
            allowed = true
            limitReached = false
            primary = nil
            secondary = nil
        }

        var additionalLimits: [UsageLimit] = []
        if let rawAdditionalLimits = JSONSupport.array(root["additional_rate_limits"]) {
            for (index, item) in rawAdditionalLimits.enumerated() {
                guard let item = JSONSupport.object(item) else {
                    continue
                }
                let nested = JSONSupport.object(item["rate_limit"]) ?? item
                let additionalPrimary = parseWindow(JSONSupport.object(nested["primary_window"]))
                let additionalSecondary = parseWindow(JSONSupport.object(nested["secondary_window"]))
                if additionalPrimary != nil || additionalSecondary != nil {
                    let name = JSONSupport.string(item["limit_name"])
                    let feature = JSONSupport.string(item["metered_feature"])
                    let identity = [
                        JSONSupport.string(item["limit_id"]),
                        name,
                        feature,
                        "additional"
                    ].first { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }!
                    additionalLimits.append(
                        UsageLimit(
                            id: "\(identity)-\(index)",
                            name: name,
                            meteredFeature: feature,
                            allowed: JSONSupport.bool(nested["allowed"], default: true),
                            limitReached: JSONSupport.bool(nested["limit_reached"], default: false),
                            primary: additionalPrimary,
                            secondary: additionalSecondary
                        )
                    )
                }
            }
        }

        var fiveHour = nearest(
            in: primaryCandidates,
            target: fiveHours,
            range: 10_800 ... 28_800,
            excluding: []
        )
        var weekly = nearest(
            in: primaryCandidates,
            target: week,
            range: 432_000 ... 777_600,
            excluding: fiveHour.map { [$0.id] } ?? []
        )
        var monthly = nearest(
            in: primaryCandidates,
            target: month,
            range: monthRange,
            excluding: [fiveHour?.id, weekly?.id].compactMap { $0 }
        )

        // Go and similar plans sometimes report a single window just outside the
        // standard 5-hour / weekly / 10–45-day buckets. Keep that window visible.
        if fiveHour == nil, weekly == nil, monthly == nil,
           let leftover = longest(in: primaryCandidates, excluding: []) {
            if leftover.window.windowSeconds >= monthRange.lowerBound {
                monthly = leftover
            } else if leftover.window.windowSeconds >= 86_400 {
                weekly = leftover
            } else {
                fiveHour = leftover
            }
        }

        let resetCredits = JSONSupport.object(root["rate_limit_reset_credits"])
        let rawAvailableCount = resetCredits.map {
            JSONSupport.int($0["available_count"], default: -1)
        } ?? -1
        let usageCredits: UsageCredits? = JSONSupport.object(root["credits"]).flatMap { credits in
            guard credits.keys.contains("has_credits")
                    || credits.keys.contains("unlimited")
                    || credits.keys.contains("balance") else {
                return nil
            }
            return UsageCredits(
                hasCredits: JSONSupport.bool(credits["has_credits"], default: false),
                unlimited: JSONSupport.bool(credits["unlimited"], default: false),
                balance: {
                    guard credits.keys.contains("balance"),
                          !(credits["balance"] is NSNull) else {
                        return nil
                    }
                    return JSONSupport.string(credits["balance"])
                }()
            )
        }

        return UsageSnapshot(
            planType: planType,
            allowed: allowed,
            limitReached: limitReached,
            fiveHour: fiveHour?.window,
            weekly: weekly?.window,
            monthly: monthly?.window,
            resetCreditsAvailable: rawAvailableCount >= 0 ? rawAvailableCount : nil,
            additionalLimits: additionalLimits,
            usageCredits: usageCredits,
            fetchedAt: fetchedAt
        )
    }

    private struct Candidate {
        let id: Int
        let window: UsageWindow
    }

    private static func parseWindow(_ object: JSONSupport.Object?) -> UsageWindow? {
        guard let object,
              !(object["used_percent"] is NSNull),
              !(object["limit_window_seconds"] is NSNull),
              object.keys.contains("used_percent"),
              object.keys.contains("limit_window_seconds"),
              let used = JSONSupport.double(object["used_percent"]),
              used.isFinite
        else {
            return nil
        }

        let duration = JSONSupport.int64(object["limit_window_seconds"], default: -1)
        guard duration > 0 else {
            return nil
        }

        let resetAfter = JSONSupport.int64(object["reset_after_seconds"], default: 0)
        let resetAtSeconds = JSONSupport.int64(object["reset_at"], default: 0)
        let roundedUsed: Int
        let javaRounded = floor(used + 0.5)
        if javaRounded >= Double(Int.max) {
            roundedUsed = Int.max
        } else if javaRounded <= Double(Int.min) {
            roundedUsed = Int.min
        } else {
            roundedUsed = Int(javaRounded)
        }

        return UsageWindow(
            usedPercent: roundedUsed,
            windowSeconds: duration,
            resetAfterSeconds: resetAfter,
            resetAt: resetAtSeconds > 0
                ? Date(timeIntervalSince1970: TimeInterval(resetAtSeconds))
                : nil
        )
    }

    private static func nearest(
        in candidates: [Candidate],
        target: Int64,
        range: ClosedRange<Int64>,
        excluding excludedIDs: [Int]
    ) -> Candidate? {
        var best: Candidate?
        var bestDistance = Int64.max

        for candidate in candidates
        where !excludedIDs.contains(candidate.id) && range.contains(candidate.window.windowSeconds) {
            let distance = absoluteDifference(candidate.window.windowSeconds, target)
            if distance < bestDistance {
                best = candidate
                bestDistance = distance
            }
        }
        return best
    }

    private static func longest(
        in candidates: [Candidate],
        excluding excludedIDs: [Int]
    ) -> Candidate? {
        candidates
            .filter { !excludedIDs.contains($0.id) }
            .max { $0.window.windowSeconds < $1.window.windowSeconds }
    }

    private static func absoluteDifference(_ lhs: Int64, _ rhs: Int64) -> Int64 {
        lhs >= rhs ? lhs - rhs : rhs - lhs
    }
}

File: android/app/src/main/java/dev/scorpion7slayer/modelsmeter/UsageParser.java
package dev.scorpion7slayer.modelsmeter;

import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/* JADX INFO: loaded from: classes.dex */
public final class UsageParser {
    private static final long FIVE_HOURS = 18000;
    private static final long WEEK = 604800;
    private static final long MONTH = 2592000;
    // Free-tier accounts report a single ~30-day Codex window; accept 10-45 days so calendar
    // months and drifting billing periods still classify while staying clear of the weekly
    // window's 9-day ceiling.
    private static final long MONTH_MIN = 864000;
    private static final long MONTH_MAX = 3888000;

    private UsageParser() {
    }

    public static UsageSnapshot parse(String str, long j) throws JSONException {
        boolean z;
        boolean z2;
        UsageWindow usageWindow;
        JSONObject jSONObject = new JSONObject(str);
        String strOptString = jSONObject.optString("plan_type", "");
        JSONObject jSONObjectNullableObject = nullableObject(jSONObject, "rate_limit");
        ArrayList arrayList = new ArrayList();
        ArrayList<UsageLimit> additionalLimits = new ArrayList<>();
        UsageWindow usageWindowFromJson = null;
        if (jSONObjectNullableObject == null) {
            z = true;
            z2 = false;
            usageWindow = null;
        } else {
            boolean zOptBoolean = jSONObjectNullableObject.optBoolean("allowed", true);
            boolean zOptBoolean2 = jSONObjectNullableObject.optBoolean("limit_reached", false);
            UsageWindow usageWindowFromJson2 = UsageWindow.fromJson(nullableObject(jSONObjectNullableObject, "primary_window"));
            usageWindowFromJson = UsageWindow.fromJson(nullableObject(jSONObjectNullableObject, "secondary_window"));
            if (usageWindowFromJson2 != null) {
                arrayList.add(usageWindowFromJson2);
            }
            if (usageWindowFromJson != null) {
                arrayList.add(usageWindowFromJson);
            }
            z = zOptBoolean;
            z2 = zOptBoolean2;
            usageWindow = usageWindowFromJson2;
        }
        JSONArray jSONArrayOptJSONArray = jSONObject.optJSONArray("additional_rate_limits");
        if (jSONArrayOptJSONArray != null) {
            for (int i = 0; i < jSONArrayOptJSONArray.length(); i++) {
                JSONObject jSONObjectOptJSONObject = jSONArrayOptJSONArray.optJSONObject(i);
                if (jSONObjectOptJSONObject != null) {
                    JSONObject jSONObjectNullableObject2 = nullableObject(jSONObjectOptJSONObject, "rate_limit");
                    if (jSONObjectNullableObject2 == null) {
                        jSONObjectNullableObject2 = jSONObjectOptJSONObject;
                    }
                    UsageWindow usageWindowFromJson3 = UsageWindow.fromJson(nullableObject(jSONObjectNullableObject2, "primary_window"));
                    UsageWindow usageWindowFromJson4 = UsageWindow.fromJson(nullableObject(jSONObjectNullableObject2, "secondary_window"));
                    if (usageWindowFromJson3 != null || usageWindowFromJson4 != null) {
                        String name = jSONObjectOptJSONObject.optString("limit_name", "");
                        String feature = jSONObjectOptJSONObject.optString("metered_feature", "");
                        String id = firstNonEmpty(
                                jSONObjectOptJSONObject.optString("limit_id", ""),
                                name, feature, "additional");
                        additionalLimits.add(new UsageLimit(
                                id + "-" + i,
                                name,
                                feature,
                                jSONObjectNullableObject2.optBoolean("allowed", true),
                                jSONObjectNullableObject2.optBoolean("limit_reached", false),
                                usageWindowFromJson3,
                                usageWindowFromJson4));
                    }
                }
            }
        }
        UsageWindow usageWindowNearest = nearest(arrayList, FIVE_HOURS, 10800L, 28800L);
        UsageWindow usageWindowNearestExcluding = nearestExcluding(arrayList, WEEK, 432000L, 777600L, usageWindowNearest);
        UsageWindow usageWindowMonthly = nearestExcluding(arrayList, MONTH, MONTH_MIN, MONTH_MAX,
                usageWindowNearest, usageWindowNearestExcluding);
        JSONObject jSONObjectNullableObject3 = nullableObject(jSONObject, "rate_limit_reset_credits");
        UsageCredits usageCredits = UsageCredits.fromJson(nullableObject(jSONObject, "credits"));
        return new UsageSnapshot(
                strOptString,
                z,
                z2,
                usageWindowNearest,
                usageWindowNearestExcluding,
                usageWindowMonthly,
                additionalLimits,
                usageCredits,
                jSONObjectNullableObject3 == null
                        ? -1 : jSONObjectNullableObject3.optInt("available_count", -1),
                j);
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) {
            if (value != null && !value.trim().isEmpty()) {
                return value.trim();
            }
        }
        return "";
    }

    private static JSONObject nullableObject(JSONObject jSONObject, String str) {
        if (jSONObject == null || jSONObject.isNull(str)) {
            return null;
        }
        return jSONObject.optJSONObject(str);
    }

    private static UsageWindow nearest(List<UsageWindow> list, long j, long j2, long j3) {
        UsageWindow usageWindow;
        long j4;
        UsageWindow usageWindow2 = null;
        long j5 = Long.MAX_VALUE;
        for (UsageWindow usageWindow3 : list) {
            if (usageWindow3.windowSeconds < j2 || usageWindow3.windowSeconds > j3) {
                long j6 = j5;
                usageWindow = usageWindow2;
                j4 = j6;
            } else {
                long jAbs = Math.abs(usageWindow3.windowSeconds - j);
                if (jAbs < j5) {
                    usageWindow = usageWindow3;
                    j4 = jAbs;
                } else {
                    long j7 = j5;
                    usageWindow = usageWindow2;
                    j4 = j7;
                }
            }
            usageWindow2 = usageWindow;
            j5 = j4;
        }
        return usageWindow2;
    }

    private static UsageWindow nearestExcluding(List<UsageWindow> list, long target,
            long minimumSeconds, long maximumSeconds, UsageWindow... excluded) {
        UsageWindow best = null;
        long bestDistance = Long.MAX_VALUE;
        for (UsageWindow candidate : list) {
            if (isExcluded(candidate, excluded)
                    || candidate.windowSeconds < minimumSeconds
                    || candidate.windowSeconds > maximumSeconds) {
                continue;
            }
            long distance = Math.abs(candidate.windowSeconds - target);
            if (distance < bestDistance) {
                best = candidate;
                bestDistance = distance;
            }
        }
        return best;
    }

    private static boolean isExcluded(UsageWindow candidate, UsageWindow[] excluded) {
        for (UsageWindow window : excluded) {
            if (candidate == window) {
                return true;
            }
        }
        return false;
    }

}
